Fix bounds derived from legacy package sets, bugfixes

MonoidMusician opened the pull request:
Fixes #531. Supersedes #567.

Takes the union of bounds generated from package sets for a particular import. Fixes several bugs we found along the way, some of which were also encountered in the previous PR.

The two PRs generated from this PR (reuploads already went to S3):
- https://github.com/purescript/registry/pull/151
- https://github.com/purescript/registry-index/pull/5

thomashoneyman commented on lib/src/Range.purs:
```diff
@@ -128,3 +129,7 @@ intersect (Range r1) (Range r2)
       , rhs: min r1.rhs r2.rhs
       }
+
+mk :: Version -> Version -> Maybe Range
```
@f-f you needed this for spago, right?

f-f replied:
Not really, see #582

thomashoneyman commented on scripts/src/PackageDeleter.purs:
```diff
@@ -138,35 +170,62 @@ main = launchAff_ do
+    Just (Metadata oldMetadata) -> do
+      let
+        publishment = case Map.lookup version oldMetadata.published, Map.lookup version oldMetadata.unpublished of
+          Just _, Just _ -> unsafeCrashWith $ "Package version was both published and unpublished: " <> formatted
```
Could we do `Except.throw` here?

thomashoneyman commented on scripts/src/Solver.purs:
```diff
@@ -0,0 +1,218 @@
+  -> Map PackageName (Map Version (Map PackageName Range))
+  -> Map PackageName (Map Version (Map PackageName (These Range Range)))
+diff = unionWithMapThese \_ -> filtered <<< these
+  do map (map This)
+  do map (map That)
+  do
+    unionWithMapThese \_ -> filtered <<< these
+      do map This
+      do map That
+      do
+        unionWithMapThese \_ -> these
+          do Just <<< This
+          do Just <<< That
+          \l r ->
+            if l == r then Nothing else Just (Both l r)
```
Well, this is something new 😆

MonoidMusician replied:
Hey, I copied `unionWithMapThese` from my project that Nate used to stress test purs-tidy, I needed to continue in the tradition ;)
https://github.com/MonoidMusician/dhall-purescript/blob/469c3e10d51a8afb90f2e231cf3e6101f50814eb/src/Dhall/Map.purs#L52-L65

thomashoneyman replied:
I'm referring to the `do`s :laughing:

f-f replied:
Yea, I don't understand this code, could we rewrite it to be more explicit? The `do`s are probably doing something here, but also make it very confusing to read

MonoidMusician replied:
The `do`s are basically bullet points, I'm using them to separate multiple arguments using indentation instead of a whole mess of parentheses.

MonoidMusician replied:
Anyways, I'm actually removing this since it isn't really relevant any more and was relying on a temporary checkout of the registry.

f-f replied:
Great!

thomashoneyman commented on scripts/src/Solver.purs:
```diff
@@ -0,0 +1,218 @@
+              let
+                deps =
+                  case Map.lookup package registry of
+                    Nothing -> unsafeCrashWith $ "Missing package: " <> PackageName.print package
```
Same here wrt `Except.throw`, since we have it available via `PublishEffects`

thomashoneyman commented on scripts/src/Solver.purs:
```diff
@@ -0,0 +1,218 @@
+module Registry.Scripts.Solver where
```
Do you mind adding a module header with a comment describing what this script does and why it exists? Do you feel there is any point at which we should remove this script in the future?

MonoidMusician replied:
I think we should keep it, it seems like it will be useful to run the solver and see what it resolves to or what the failures are. Especially because we can modify the registry index and test quick fixes locally, as opposed to a spago workflow which wants a clean registry index. And because I'm proud of my solver :D

thomashoneyman commented on app/src/App/Effect/Storage.purs:
```diff
@@ -45,6 +46,7 @@ data Storage a
   = Upload PackageName Version FilePath (Either String Unit -> a)
   | Download PackageName Version FilePath (Either String Unit -> a)
   | Delete PackageName Version (Either String Unit -> a)
+  | Query PackageName (Either String (Array String) -> a)
```
Since I believe these are package paths, of the form `"package/x.y.z.tar.gz"`, I feel like `Query` should return an `Array Version`

MonoidMusician replied:
That's correct. I didn't feel like parsing it tbh but I can if you want.

thomashoneyman replied:
I would prefer that, thank you!

thomashoneyman commented on app/src/App/Effect/Storage.purs:
```diff
@@ -183,6 +198,9 @@ handleS3 env = Cache.interpret _storageCache (Cache.handleFs env.cache) <<< case
 handleReadOnly :: forall r a. FilePath -> Storage a -> Run (LOG + AFF + EFFECT + r) a
 handleReadOnly cache = Cache.interpret _storageCache (Cache.handleFs cache) <<< case _ of
+  Query _ reply -> do
+    pure $ reply $ Left "Cannot query in read-only mode."
```
You should be able to query in read-only mode, as it does not modify the registry.

MonoidMusician replied:
It doesn't have tokens to connect to s3 in read-only mode, unless you know of a different way.

thomashoneyman replied:
Oh, yea, we have to be connected to S3 to query objects. That's...kind of lame. Well, querying certainly seems like something we would want to have available in a read-only mode, but perhaps you can just add a TODO comment here.

thomashoneyman commented on scripts/src/VerifyIntegrity.purs:
```diff
@@ -0,0 +1,187 @@
+    gitEnv =
+      { write: ReadOnly
+      , pull: Autostash
+      , repos: Git.defaultRepos { manifestIndex = Git.defaultRepos.manifestIndex { owner = "monoidmusician" }, registry = Git.defaultRepos.registry { owner = "monoidmusician" } }
```
I don't think we should hardcode these values. I think we should use default repos, and either take the overrides via the command line, or expect that if you are working on the scripts locally then you override it yourself.

MonoidMusician replied:
yeah I removed that already

thomashoneyman commented on scripts/src/VerifyIntegrity.purs:
```diff
@@ -0,0 +1,187 @@
+  interpret do
+    let
+      sorry e = Log.error e *> liftEffect (Process.exit 1)
+    allMetadata <- Except.runExcept Registry.readAllMetadata >>= either (map absurd <<< sorry) pure
```
Why not just add an `Except` interpreter that exits the process?

```purs
>>> Except.catch (\error -> Run.liftEffect (Console.log error *> Process.exit 1))
```

and then, where you're using `sorry` you can `Log.error err *> Except.fail`. For example, since you're not doing anything with the exception when reading metadata it could be:

```purs
interpret do
  allMetadata <- Registry.readAllMetadata
  allManifests <- Registry.readAllManifests
  ...
```

MonoidMusician replied:
You mean adding that globally? Ideally it wouldn't stop on each package that fails so you could see all the failures.

thomashoneyman replied:
You're saying you want the process to exit as soon as you encounter one of these exceptions?

MonoidMusician replied:
Oh, I misunderstood your suggestion. Makes sense!

thomashoneyman commented on scripts/src/VerifyIntegrity.purs:
```diff
@@ -0,0 +1,187 @@
+      [ names.right <> " missing " <> setOf f left ]
+  ]
+
+verifyPackage :: forall r. Map PackageName Metadata -> ManifestIndex -> Set String -> PackageName -> Run (EXCEPT String + LOG + r) Unit
```
Could you add a comment about the rules for verification, e.g. the metadata is the source of truth; there should be zero versions present in S3 / registry-index that are not also present in the metadata's 'published' field, and every version present in the metadata's 'published' field should also be present in s3 / the registry-index?

MonoidMusician commented:
Okie, I think I addressed all your feedback. Thanks for the review, the help in writing it, and especially the new effects! I'm liking how the effects work out, I think it's pretty clean.
